[Ui-side compositing] Fix mouse position reporting to AppKit in WebScrollerImpPairDelegateMac

nmoucht opened the pull request:
#### caec1d312c597c37515751aedbec9433978fb627
<pre>
[Ui-side compositing] Fix mouse position reporting to AppKit in WebScrollerImpPairDelegateMac
<a href="https://bugs.webkit.org/show_bug.cgi?id=255188">https://bugs.webkit.org/show_bug.cgi?id=255188</a>
rdar://107776583

Reviewed by Simon Fraser.

This patch aims to fix several issues with the way ScrollerPairMac and ScrollerMac handle
mouse events, with the intent of fixing hovering over overlay scrollbars in overflow scrolling
nodes. The first issue is that currently there isn&apos;t a good way to convert the frame relative
mouse position to a scrollbar relative mouse position. To work around this, bundle the scrollbar
relative mouse positions in the mouseMovedInContentArea event. When we know that we have updated
mouse positions from the WebProcess, return them in the various WebScrollerImpPairDelegateMac
functions. The second issue is that RemoteScrollingTree is only calling handleMouseEvent
on the root scroller. Since this mouse position was only being used as input to convertContentPoint,
it is ok to just return a zero point and use the converted mouse position from the Web Process.

* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::lastKnownMousePosition):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::lastKnownMousePosition const):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNodeDelegate.h:
(WebCore::ScrollingTreeScrollingNodeDelegate::lastKnownMousePosition const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(-[WebScrollerImpDelegateMac mouseLocationInScrollerForScrollerImp:]):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
(WebCore::ScrollerPairMac::node const):
(WebCore::ScrollerPairMac::lastKnownMousePosition const): Deleted.
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac mouseLocationInContentAreaForScrollerImpPair:]):
(-[WebScrollerImpPairDelegateMac scrollerImpPair:convertContentPoint:toScrollerImp:]):
(WebCore::ScrollerPairMac::lastKnownMousePosition const):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::lastKnownMousePosition const):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::lastKnownMousePosition):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::lastKnownMousePosition):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.h:

Canonical link: <a href="https://commits.webkit.org/262995@main">https://commits.webkit.org/262995@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bead618f98d61ba1b4695d6c5f1789ef1edaa262

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3188 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3251 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3361 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4603 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3546 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3158 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3316 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3278 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2776 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3221 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3539 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4413 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1039 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2843 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2683 "3 flakes 147 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2818 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2894 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4167 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3266 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2610 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2848 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2845 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2843 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/379 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3107 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -286,8 +287,9 @@ void ScrollingStateScrollingNode::setMouseIsOverContentArea(bool flag)
 }
 
-void ScrollingStateScrollingNode::setMouseMovedInContentArea()
+void ScrollingStateScrollingNode::setMouseMovedInContentArea(MouseLocationState mouseLocationState)
```
pass by const ref?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
```diff
@@ -45,6 +45,11 @@ struct ScrollbarHoverState {
 };
 
+struct MouseLocationState {
+    IntPoint locationRelativeToHorizontalScrollbar;
```
“locationIn” is OK

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
```diff
@@ -120,7 +125,8 @@ class ScrollingStateScrollingNode : public ScrollingStateNode {
 
-    WEBCORE_EXPORT void setMouseMovedInContentArea();
+    WEBCORE_EXPORT void setMouseMovedInContentArea(MouseLocationState);
+    MouseLocationState mouseLocationState() const { return m_mouseLocationState; }
```
Return a const&

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.h:
```diff
@@ -242,6 +242,7 @@ class ScrollingTree : public ThreadSafeRefCounted<ScrollingTree> {
 
     WEBCORE_EXPORT Seconds frameDuration();
     WEBCORE_EXPORT Seconds maxAllowableRenderingUpdateDurationForSynchronization();
+    virtual IntPoint lastKnownMousePosition() { return { }; }
```
Relative to what, though? I don’t think we can have this here unless it’s really clear what coordinate system it’s in.

nmoucht replied:
Renamed to be raw mouse position.

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
```diff
@@ -132,6 +132,7 @@ class WEBCORE_EXPORT ScrollingTreeScrollingNode : public ScrollingTreeNode {
     OverscrollBehavior verticalOverscrollBehavior() const { return m_scrollableAreaParameters.verticalOverscrollBehavior; }
     
     virtual String scrollbarStateForOrientation(ScrollbarOrientation) const { return ""_s; }
+    IntPoint lastKnownMousePosition() const;
```
Again, confusion about coordinate systems.

nmoucht replied:
Renamed to be raw mouse position.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -369,6 +370,8 @@ - (void)invalidate
 FloatPoint ScrollerMac::convertFromContent(const FloatPoint& point) const
 {
+    if (m_pair.mouseInContentArea())
+        return m_lastKnownMousePosition;
```
This doesn’t make sense. The function is asked to convert `point` into the scrollbar’s coordinate system. It’s not right to just return `m_lastKnownMousePosition` here.

nmoucht replied:
I think the confusion here is ScrollerPairMac has m_lastKnownMousePosition which is the raw mouse position, while each ScrollerMac has m_lastKnownMousePosition which is the raw mouse position relative to that scroller. I renamed each of these to be more clear. ScrollerMac::convertFromContent is called with the raw mouse position so if m_pair.mouseInContentArea() is true we know the converted mouse position is updated by the Web Process.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -83,8 +84,10 @@ - (NSPoint)mouseLocationInContentAreaForScrollerImpPair:(NSScrollerImpPair *)scr
     if (!_scrollerPair)
         return NSZeroPoint;
+    auto location = _scrollerPair->lastKnownMousePosition();
+    WTFLogAlways("ScrollerMac::mouseLocationInContentAreaForScrollerImpPair: node:%d {%d,%d}", _scrollerPair->node().scrollingNodeID(), location.x(), location.y());
```
Logging

nmoucht replied:
Fixed.

smfr reviewed (changes requested):
We need to think of a cleaner way to do this.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -286,8 +287,9 @@ void ScrollingStateScrollingNode::setMouseIsOverContentArea(bool flag)
-void ScrollingStateScrollingNode::setMouseMovedInContentArea()
+void ScrollingStateScrollingNode::setMouseMovedInContentArea(const MouseLocationState& mouseLocationState)
 {
+    m_mouseLocationState = mouseLocationState;
```
This is one place we could optimize and avoid scrolling tree transactions on every mouse move.

If we only care about mouse moves when over the scrollbar, we should have `AsyncScrollingCoordinator::setMouseMovedInContentArea` always use some placeholder value (-1, -1) when the mouse is not over a scrollbar, and then this function could early return if m_mouseLocationState is unchanged.

nmoucht replied:
This is a good idea, but the mouseMovedInContent area event also has the affect of keeping the scrollbar visible while those events are still coming in. I think this would be a good backup if we discover that the extra scrolling tree transactions are a big perf hit.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -105,12 +102,14 @@ class ScrollerPairMac : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<S
     void mouseEnteredContentArea();
     void mouseExitedContentArea();
-    void mouseMovedInContentArea();
+    void mouseMovedInContentArea(MouseLocationState);
```
`const MouseLocationState&`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -409,6 +405,15 @@ - (void)invalidate
 
+IntPoint ScrollerMac::lastKnownMousePositionInScrollbar() const
+{
+    // When we dont have an update from the Web Process, return
```
This comment is only relevant to UI-side compositng, but you have avoid breaking non-UI-side composting too!

nmoucht replied:
ScrollerPairMac/ScrollerMac are meant to be UI-side compositing only classes. This code is handled by the delegate in ScrollbarsContollerMac when in non-UI-side composting mode.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -81,15 +81,15 @@ - (BOOL)inLiveResizeForScrollerImpPair:(NSScrollerImpPair *)scrollerImpPair
-
-    return _scrollerPair->lastKnownMousePosition();
+    // This location is only used when calling mouseLocationInScrollerForScrollerImp,
+    // where we will use the converted mouse position from the Web Process
```
This comment also is only relevant to UI-side compositing.

nmoucht replied:
See above.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
```diff
@@ -45,6 +45,11 @@ struct ScrollbarHoverState {
     friend bool operator==(ScrollbarHoverState const& lhs, ScrollbarHoverState const& rhs) = default;
 };
 
+struct MouseLocationState {
```
Should this be merged with `ScrollbarHoverState`?

nmoucht replied:
Since these states are not coupled I think it maks sense to keep them seperate. If we end up coupling them like you suggested above, then I think that would be reasonable.

smfr reviewed (approved):
r=me but please make sure this doesn’t break non-UI-side compositing or WebKit1.

nmoucht committed:
[Ui-side compositing] Fix mouse position reporting to AppKit in WebScrollerImpPairDelegateMac
https://bugs.webkit.org/show_bug.cgi?id=255188
rdar://107776583

Reviewed by Simon Fraser.

This patch aims to fix several issues with the way ScrollerPairMac and ScrollerMac handle
mouse events, with the intent of fixing hovering over overlay scrollbars in overflow scrolling
nodes. The first issue is that currently there isn't a good way to convert the frame relative
mouse position to a scrollbar relative mouse position. To work around this, bundle the scrollbar
relative mouse positions in the mouseMovedInContentArea event. When we know that we have updated
mouse positions from the WebProcess, return them in the various WebScrollerImpPairDelegateMac
functions. The second issue is that RemoteScrollingTree is only calling handleMouseEvent
on the root scroller. Since this mouse position was only being used as input to convertContentPoint,
it is ok to just return a zero point and use the converted mouse position from the Web Process.

* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::lastKnownMousePosition):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::lastKnownMousePosition const):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNodeDelegate.h:
(WebCore::ScrollingTreeScrollingNodeDelegate::lastKnownMousePosition const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(-[WebScrollerImpDelegateMac mouseLocationInScrollerForScrollerImp:]):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
(WebCore::ScrollerPairMac::node const):
(WebCore::ScrollerPairMac::lastKnownMousePosition const): Deleted.
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac mouseLocationInContentAreaForScrollerImpPair:]):
(-[WebScrollerImpPairDelegateMac scrollerImpPair:convertContentPoint:toScrollerImp:]):
(WebCore::ScrollerPairMac::lastKnownMousePosition const):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::lastKnownMousePosition const):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::lastKnownMousePosition):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::lastKnownMousePosition):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.h:

Canonical link: https://commits.webkit.org/262995@main
